src/peerbook/libp2p_peerbook.erl

```diff
@@ -686,27 +693,38 @@ notify_peers(State=#state{notify_peers=NotifyPeers}) when map_size(NotifyPeers)
 notify_peers(State=#state{notify_peers=NotifyPeers, notify_group=NotifyGroup,
                           gossip_group=GossipGroup, tid=TID}) ->
     %% Notify to local interested parties
-    PeerList = maps:values(NotifyPeers),
-    [Pid ! {new_peers, PeerList} || Pid <- ?PG_MEMBERS(NotifyGroup)],
+    PeerList0 = maps:values(NotifyPeers),
+    [Pid ! {new_peers, PeerList0} || Pid <- ?PG_MEMBERS(NotifyGroup)],
+
+    %% pre-encode all the peers so we avoid re-encoding them every time we subsample the list
+    %% these can just be joined/prefixed with 10,<length> to simulate the list wrapping
+    %% because these came out of a map, we assume the ordering is pretty random
+    PeerList = [ begin
+                     Encoded = libp2p_peer:encode(P),
```

Comment:
this is wildly cute, a bit brittle, and would it not be better of in libp2p_peer:encode_list to begin with?

also, is this loosing the fact that we strip out metadata from the peer before encoding it?

Reply:
Yes, these are valid points.